src/xbot/common/subsystems/pose/BasePoseSubsystem.java

```diff
@@ -0,0 +1,185 @@
+import xbot.common.math.ContiguousHeading;
+import xbot.common.math.XYPair;
+import xbot.common.properties.BooleanProperty;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+
+public abstract class BasePoseSubsystem extends BaseSubsystem implements PeriodicDataSource {
+
+    public final XGyro imu;
+    
+    private final DoubleProperty leftDriveDistance;
+    private final DoubleProperty rightDriveDistance;
+    
+    private final DoubleProperty totalDistanceX;
+    private final DoubleProperty totalDistanceY;
+    
+    private ContiguousHeading currentHeading;
+    private final DoubleProperty currentHeadingProp;
+    
+    private ContiguousHeading lastImuHeading;
+    
+    // These are two common robot starting positions - kept here as convenient shorthand.
+    public static final double FACING_AWAY_FROM_DRIVERS = 90;
+    public static final double FACING_TOWARDS_DRIVERS = -90;
+    
+    private final DoubleProperty currentPitch;
+    private final DoubleProperty currentRoll;
+    
+    private final DoubleProperty inherentRioPitch;
+    private final DoubleProperty inherentRioRoll;
+    
+    private double previousLeftDistance;
+    private double previousRightDistance;
+    
+    private BooleanProperty rioRotated;
+    
+    public BasePoseSubsystem(WPIFactory factory, XPropertyManager propManager) {
+        log.info("Creating");
+        imu = factory.getGyro(ImuType.navX);
+        
+        currentHeadingProp = propManager.createEphemeralProperty("CurrentHeading", 0.0);
+        // Right when the system is initialized, we need to have the old value be
+        // the same as the current value, to avoid any sudden changes later
+        lastImuHeading = imu.getYaw();
+        currentHeading = new ContiguousHeading(FACING_AWAY_FROM_DRIVERS);
+        
+        currentPitch = propManager.createEphemeralProperty("CurrentPitch", 0.0);
+        currentRoll = propManager.createEphemeralProperty("CurrentRoll", 0.0);
+        
+        leftDriveDistance = propManager.createEphemeralProperty("LeftDriveDistance", 0.0);
+        rightDriveDistance = propManager.createEphemeralProperty("RightDriveDistance", 0.0);
+        
+        totalDistanceX = propManager.createEphemeralProperty("TotalDistanceX", 0.0);
+        totalDistanceY = propManager.createEphemeralProperty("TotalDistanceY", 0.0);
+        
+        rioRotated = propManager.createPersistentProperty("RioRotated", false);
+        inherentRioPitch = propManager.createPersistentProperty("InherentRioPitch", 0.0);
+        inherentRioRoll = propManager.createPersistentProperty("InherentRioRoll", 0.0);
+    }
+    
+    private void updateCurrentHeading() {
+        // Old heading - current heading gets the delta heading        
+        double imuDeltaYaw = lastImuHeading.difference(imu.getYaw());
+
+        // add the delta to our current
+        currentHeading.shiftValue(imuDeltaYaw);
```

Comment:
The other advantage of this approach is that it lets the PoseSubsystem work with any given initial raw gyro value - no matter what value the gyro is at when the robot boots, since we're only looking at deltas in the measured value, we always start in a known config.